jsonpb/jsonpb.go

```diff
@@ -644,7 +649,105 @@ func (u *Unmarshaler) UnmarshalNext(dec *json.Decoder, pb proto.Message) error {
 // permutations of the related Marshaler.
 func (u *Unmarshaler) Unmarshal(r io.Reader, pb proto.Message) error {
 	dec := json.NewDecoder(r)
-	return u.UnmarshalNext(dec, pb)
+	if err := u.UnmarshalNext(dec, pb); err != nil {
+		return err
+	}
+
+	return checkRequiredFields(pb)
+}
+
+// checkRequiredFields returns an error if any required field in the given proto message is not set.
+// This function is called from Unmarshal and assumes certain actions are done ahead.  While
+// required fields only exist in a proto2 message, a proto3 message can contain proto2 message(s).
+func checkRequiredFields(pb proto.Message) error {
+	// Most well-known type messages do not contain required fields.  The "Any" type may contain
+	// a message that has required fields.  When an Any message is being unmarshaled, the code will
+	// have invoked proto.Marshal on the embedded message to store the serialized message in
+	// Any.Value field, and that should have returned an error if a required field is not set.
+	// Hence skipping well-known types here.
+	if _, ok := pb.(wkt); ok {
+		return nil
+	}
+
+	v := reflect.ValueOf(pb).Elem()
+	for i := 0; i < v.NumField(); i++ {
+		field := v.Field(i)
+		sfield := v.Type().Field(i)
+		if strings.HasPrefix(sfield.Name, "XXX_") {
+			continue
+		}
+
+		// A oneof field is an interface implemented by wrapper structs containing the actual oneof
+		// fields.  Field is an interface containing &T{real_value}.
+		if sfield.Tag.Get("protobuf_oneof") != "" {
+			v := field.Elem().Elem() // interface -> *T -> T
+			field = v.Field(0)
+			sfield = v.Type().Field(0)
+		}
+
+		var prop proto.Properties
+		prop.Init(sfield.Type, sfield.Name, sfield.Tag.Get("protobuf"), &sfield)
+
+		switch field.Kind() {
+		case reflect.Map:
+			if field.IsNil() {
+				continue
+			}
+			// Check each map value.
+			keys := field.MapKeys()
+			for _, k := range keys {
+				v := field.MapIndex(k)
+				if err := checkRequiredFieldsInValue(v); err != nil {
+					return err
+				}
+			}
+		case reflect.Slice:
```

Comment:
The "bytes" scalar for proto will fall under this case and will need to be specially handled.
IIRC, `[]byte{}` means that a zero-length bytes is set, while `[]byte(nil)` means that it was never set.

Reply:
Good catch, I forgot about that.  Added handling of non-repeated slice.

Yes, I'm going with []byte{} to mean it is set.